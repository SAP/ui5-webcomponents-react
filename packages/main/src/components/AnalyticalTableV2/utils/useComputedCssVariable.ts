import type { RefObject } from 'react';
import { useEffect, useState } from 'react';

export function useComputedCssVariable<T extends HTMLElement>(ref: RefObject<T>, variableName: string): number | null {
  const [value, setValue] = useState<number | null>(null);

  useEffect(() => {
    if (ref.current) {
      const computedStyles = getComputedStyle(ref.current);
      const variableValue = computedStyles.getPropertyValue(variableName).trim();

      // Attempt to parse the value as a number
      const parsedValue = parseFloat(variableValue);
      setValue(isNaN(parsedValue) ? null : parsedValue);
    }
  }, [variableName, ref]);

  return value;
}
