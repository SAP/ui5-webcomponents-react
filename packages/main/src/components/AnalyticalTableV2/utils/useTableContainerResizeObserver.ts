import type { RefObject } from 'react';
import { useState, useEffect } from 'react';

export const useTableContainerResizeObserver = (tableContainerRef: RefObject<HTMLDivElement>) => {
  const [tableWidth, setTableWidth] = useState(0);
  const [horizontalScrollbarHeight, setHorizontalScrollbarHeight] = useState(0);

  useEffect(() => {
    const tableContainer = tableContainerRef.current;

    const resizeObserver = new ResizeObserver((entries) => {
      for (const entry of entries) {
        if (entry.borderBoxSize) {
          const { borderBoxSize, contentBoxSize } = entry;
          const borderBoxHeight = borderBoxSize[0].blockSize;
          const contentBoxHeight = contentBoxSize[0].blockSize;
          if (borderBoxHeight > contentBoxHeight) {
            setHorizontalScrollbarHeight(borderBoxHeight - contentBoxHeight);
          } else {
            setHorizontalScrollbarHeight(0);
          }

          const borderBoxWidth = borderBoxSize[0].inlineSize;
          setTableWidth(borderBoxWidth);
        }
      }
    });

    if (tableContainer) {
      resizeObserver.observe(tableContainer, { box: 'border-box' });
    }

    return () => {
      if (tableContainer) {
        resizeObserver.unobserve(tableContainer);
      }
    };
  }, [tableContainerRef]);

  return { tableWidth, horizontalScrollbarHeight };
};
