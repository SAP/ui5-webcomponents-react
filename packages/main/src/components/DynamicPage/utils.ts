export const DynamicPageCssVariables = {
  headerDisplay: '--_ui5wcr_DynamicPage_header_display',
  titleFontSize: '--_ui5wcr_DynamicPage_title_fontsize'
};
