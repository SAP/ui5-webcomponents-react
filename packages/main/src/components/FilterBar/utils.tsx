import { cloneElement } from 'react';

const inputTagNames = new Set([
  'UI5-COMBOBOX',
  'UI5-DATE-PICKER',
  'UI5-DATERANGE-PICKER',
  'UI5-DATETIME-PICKER',
  'UI5-DURATION-PICKER',
  'UI5-INPUT',
  'UI5-MULTI-INPUT',
  'UI5-RATING-INDICATOR',
  'UI5-STEP-INPUT',
  'UI5-TIME-PICKER'
]);

export const filterValue = (ref, child) => {
  const tagName = ref.tagName;
  let filterItemProps = {};
  if (inputTagNames.has(tagName)) {
    filterItemProps = { value: ref.value };
  }
  if (tagName === 'UI5-SELECT' || tagName === 'UI5-MULTI-COMBOBOX') {
    const selectedIndices = Array.from(ref.children as HTMLCollectionOf<any>)
      .map((item, index) => (item.selected ? index : false))
      .filter((el) => el !== false);
    const selectedIndicesSet = new Set(selectedIndices);
    const options = child.props.children.props.children?.map((item, index) => {
      if (selectedIndicesSet.has(index)) {
        return cloneElement(item, { selected: true });
      }
      return cloneElement(item, { selected: false });
    });
    filterItemProps = { children: options };
  }
  if (tagName === 'UI5-SWITCH' || tagName === 'UI5-CHECKBOX') {
    filterItemProps = { checked: ref.checked };
  }
  return filterItemProps;
};

export const renderSearchWithValue = (renderSearchElement, searchValue) => {
  return cloneElement(renderSearchElement, { value: searchValue ?? renderSearchElement?.props?.value });
};
