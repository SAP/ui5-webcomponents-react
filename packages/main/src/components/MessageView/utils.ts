import iconAlert from '@ui5/webcomponents-icons/dist/alert.js';
import iconError from '@ui5/webcomponents-icons/dist/error.js';
import iconInformation from '@ui5/webcomponents-icons/dist/information.js';
import iconSysEnter from '@ui5/webcomponents-icons/dist/sys-enter-2.js';
import { ValueState } from '../../enums/ValueState';

export const getIconNameForType = (type: ValueState | keyof typeof ValueState): string => {
  switch (type) {
    case ValueState.Error:
    case 'Error':
      return iconError;
    case ValueState.Success:
    case 'Success':
      return iconSysEnter;
    case ValueState.Warning:
    case 'Warning':
      return iconAlert;
    default:
      return iconInformation;
  }
};
