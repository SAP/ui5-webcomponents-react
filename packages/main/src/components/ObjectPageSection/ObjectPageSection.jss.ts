import { ThemingParameters } from '@ui5/webcomponents-react-base';
import { CustomThemingParameters } from '../../themes/CustomVariables';

const styles = {
  header: {
    borderTop: CustomThemingParameters.ObjectPageSectionBorderTop,
    boxSizing: 'border-box',
    height: '2.25rem'
  },
  title: {
    height: '2.25rem',
    lineHeight: '2.25rem',
    color: ThemingParameters.sapGroup_TitleTextColor,
    fontSize: ThemingParameters.sapFontHeader4Size,
    display: 'inline-block',
    maxWidth: '100%',
    overflow: 'hidden',
    whiteSpace: 'nowrap',
    textOverflow: 'ellipsis',
    fontFamily: CustomThemingParameters.ObjectPageSectionTitleFontFamily,
    textDecoration: CustomThemingParameters.ObjectPageSectionTitleTextDecoration
  },
  uppercase: {
    textTransform: 'uppercase'
  },
  sectionContent: {
    whiteSpace: 'normal'
  },
  sectionContentInner: {
    paddingTop: '1rem',
    paddingBottom: '2rem',
    fontFamily: ThemingParameters.sapFontFamily
  }
};

export default styles;
