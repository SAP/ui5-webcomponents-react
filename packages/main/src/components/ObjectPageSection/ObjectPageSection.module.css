.section [data-component-name='ObjectPageSubSection']:not(:first-child) {
  padding-block-start: 0.5rem;
}

.headerContainer {
  padding-block: 0.5rem;
  color: var(--sapGroup_TitleTextColor);
  font-family: var(--sapFontFamily);
}

.titleContainer {
  border-block-start: var(--_ui5wcr_ObjectPage_SectionBorderTop);
  box-sizing: border-box;
  height: 2.25rem;
}

.title {
  height: var(--_ui5wcr_ObjectPage_SectionTitleHeight);
  line-height: var(--_ui5wcr_ObjectPage_SectionTitleLineHeight);
  color: var(--sapGroup_TitleTextColor);
  font-size: var(--sapFontHeader4Size);
  display: inline-block;
  max-width: 100%;
  overflow: hidden;
  white-space: nowrap;
  text-overflow: ellipsis;
  font-family: var(--_ui5wcr_ObjectPage_SectionTitleFontFamily);
  text-decoration: var(--_ui5wcr_ObjectPage_SectionTitleTextDecoration);
}

.uppercase {
  text-transform: uppercase;
}

.wrap {
  .titleContainer {
    height: unset;
  }
  .title {
    white-space: normal;
    height: unset;
    line-height: normal;
  }
}

.sectionContent {
  white-space: normal;
  height: 100%;
}

.sectionContentInner {
  padding-block: 0.5rem;
  font-family: var(--sapFontFamily);
  height: 100%;
  box-sizing: border-box;
}
