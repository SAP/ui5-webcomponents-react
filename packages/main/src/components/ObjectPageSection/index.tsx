import { useConsolidatedRef } from '@ui5/webcomponents-react-base/dist/useConsolidatedRef';
import { usePassThroughHtmlProps } from '@ui5/webcomponents-react-base/dist/usePassThroughHtmlProps';
import React, { forwardRef, ReactNode, ReactNodeArray, RefObject } from 'react';
import { createUseStyles } from 'react-jss';
import { CommonProps } from '@ui5/webcomponents-react/interfaces/CommonProps';
import { EmptyIdPropException } from '../ObjectPage/EmptyIdPropException';
import styles from './ObjectPageSection.jss';
import clsx from 'clsx';

export interface ObjectPageSectionPropTypes extends CommonProps {
  /**
   * Defines the title of the `ObjectPageSection`.
   */
  titleText?: string;
  /**
   * Defines the ID of the `ObjectPageSection`.<br />
   * __Note:__ The `id` is taken into account when the section selection changes.
   */
  id: string;
  /**
   * Defines whether the title is always displayed in uppercase.
   */
  titleTextUppercase?: boolean;
  /**
   * Defines the content of the `ObjectPageSection`.
   */
  children: ReactNode | ReactNodeArray;
}

const useStyles = createUseStyles(styles, { name: 'ObjectPageSection' });
/**
 * Top-level information container of an `ObjectPage`.
 */
const ObjectPageSection = forwardRef((props: ObjectPageSectionPropTypes, ref: RefObject<HTMLElement>) => {
  const { titleText, id, children, titleTextUppercase, className, style, tooltip } = props;
  const classes = useStyles();

  if (!id) {
    throw new EmptyIdPropException('ObjectPageSection requires a unique ID property!');
  }

  const sectionRef: RefObject<HTMLElement> = useConsolidatedRef(ref);
  const htmlId = `ObjectPageSection-${id}`;

  const titleClasses = clsx(classes.title, titleTextUppercase && classes.uppercase);

  const passThroughProps = usePassThroughHtmlProps(props, ['id']);
  return (
    <section
      ref={sectionRef}
      role="region"
      className={className}
      style={style}
      title={tooltip}
      {...passThroughProps}
      id={htmlId}
      data-component-name="ObjectPageSection"
    >
      <div role="heading" aria-level={3} className={classes.header} data-component-name="ObjectPageSectionTitleText">
        <div className={titleClasses}>{titleText}</div>
      </div>
      {/* TODO Check for subsections as they should win over the children */}
      <div className={classes.sectionContent}>
        <div className={classes.sectionContentInner} data-component-name="ObjectPageSectionContent">
          {children}
        </div>
      </div>
    </section>
  );
});

ObjectPageSection.defaultProps = {
  titleText: '',
  titleTextUppercase: true
};

ObjectPageSection.displayName = 'ObjectPageSection';

export { ObjectPageSection };
