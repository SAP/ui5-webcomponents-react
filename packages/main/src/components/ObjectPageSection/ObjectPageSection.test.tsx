import { render } from '@shared/tests';
import { createPassThroughPropsTest } from '@shared/tests/utils';
import { ObjectPageSection } from '@ui5/webcomponents-react/dist/ObjectPageSection';
import React from 'react';

describe('ObjectPageSection', () => {
  test('Renders with children', () => {
    const { asFragment } = render(
      <ObjectPageSection id={'1'} titleText="Test" titleTextUppercase>
        This is my Text
      </ObjectPageSection>
    );
    expect(asFragment()).toMatchSnapshot();
  });

  test('ObjectPage w/ lowercase title', () => {
    const { asFragment } = render(
      <ObjectPageSection id={'1'} titleText="Test" titleTextUppercase={false}>
        This is my Text
      </ObjectPageSection>
    );
    expect(asFragment()).toMatchSnapshot();
  });

  test('Empty Id Prop', () => {
    const renderer = () => render(<ObjectPageSection />);
    expect(renderer).toThrow();
  });

  createPassThroughPropsTest(ObjectPageSection);
});
