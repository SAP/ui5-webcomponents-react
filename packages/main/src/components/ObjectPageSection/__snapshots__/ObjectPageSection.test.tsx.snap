// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ObjectPageSection ObjectPage w/ lowercase title 1`] = `
<DocumentFragment>
  <section
    data-component-name="ObjectPageSection"
    id="ObjectPageSection-1"
    role="region"
  >
    <div
      aria-level="3"
      class="ObjectPageSection-header"
      data-component-name="ObjectPageSectionHeading"
      role="heading"
    >
      <div
        class="ObjectPageSection-title"
      >
        Test
      </div>
    </div>
    <div
      class="ObjectPageSection-sectionContent"
    >
      <div
        class="ObjectPageSection-sectionContentInner"
        data-component-name="ObjectPageSectionContent"
      >
        This is my Text
      </div>
    </div>
  </section>
</DocumentFragment>
`;

exports[`ObjectPageSection Renders with children 1`] = `
<DocumentFragment>
  <section
    data-component-name="ObjectPageSection"
    id="ObjectPageSection-1"
    role="region"
  >
    <div
      aria-level="3"
      class="ObjectPageSection-header"
      data-component-name="ObjectPageSectionHeading"
      role="heading"
    >
      <div
        class="ObjectPageSection-title ObjectPageSection-uppercase"
      >
        Test
      </div>
    </div>
    <div
      class="ObjectPageSection-sectionContent"
    >
      <div
        class="ObjectPageSection-sectionContentInner"
        data-component-name="ObjectPageSectionContent"
      >
        This is my Text
      </div>
    </div>
  </section>
</DocumentFragment>
`;
